src/draw/lv_image_decoder.c

```diff
@@ -350,6 +351,156 @@ static lv_image_decoder_built_in_data_t * get_decoder_data(lv_image_decoder_dsc_
+    return size;
+}
+
+static lv_result_t decode_indexed(lv_image_decoder_t * decoder, lv_image_decoder_dsc_t * dsc)
+{
+    lv_result_t res;
+    lv_image_decoder_built_in_data_t * decoder_data = dsc->user_data;
+    lv_fs_file_t *f = &decoder_data->f;
+    lv_color_format_t cf = dsc->header.cf;
+
+    /*read palette for indexed image*/
+    uint32_t palette_len = sizeof(lv_color32_t) * LV_COLOR_INDEXED_PALETTE_SIZE(cf);
+    lv_color32_t * palette = lv_malloc(palette_len);
+    LV_ASSERT_MALLOC(palette);
+    if(palette == NULL) {
+        LV_LOG_ERROR("out of memory");
+        return LV_RESULT_INVALID;
+    }
+
+    uint32_t rn;
+    res = fs_read_file_at(f, sizeof(lv_image_header_t), (uint8_t *)palette, palette_len, &rn);
+    if(res != LV_FS_RES_OK || rn != palette_len) {
+        LV_LOG_WARN("Built-in image decoder can't read the palette");
+        lv_free(palette);
+        return LV_RESULT_INVALID;
+    }
+
+    dsc->palette = palette;
+    dsc->palette_size = LV_COLOR_INDEXED_PALETTE_SIZE(cf);
+
+    decoder_data->palette = palette; /*Free decoder data on close*/
+
+#if LV_BIN_DECODER_RAM_LOAD
+    uint32_t stride = dsc->header.stride;
+    uint8_t * file_buf = lv_draw_buf_malloc(stride * dsc->header.h, cf);
+    LV_ASSERT_MALLOC(file_buf);
+    if (file_buf == NULL) {
+        LV_LOG_ERROR("no memory for file read.\n");
+        return LV_RESULT_INVALID;
+    }
+
+    uint32_t data_len = get_filesize(f);
+    uint32_t data_offset = sizeof(lv_image_header_t) + palette_len;
+    data_len -= data_offset;
+    res = fs_read_file_at(f, data_offset, (uint8_t *)file_buf, data_len, &rn);
+    if(res != LV_FS_RES_OK || rn != data_len) {
+        LV_LOG_WARN("Built-in image decoder can't read the palette");
+        lv_free(palette);
+        lv_draw_buf_free(file_buf);
+        return LV_RESULT_INVALID;
+    }
+
+    /*Convert to ARGB8888, since sw renderer cannot render it directly even it's in RAM*/
+    stride = lv_draw_buf_width_to_stride(dsc->header.w, LV_COLOR_FORMAT_ARGB8888);
+    uint8_t * img_data = lv_draw_buf_malloc(stride * dsc->header.h, cf);
+    if (img_data == NULL) {
+        LV_LOG_ERROR("no memory for indexed image.\n");
+        lv_free(palette);
+        lv_draw_buf_free(file_buf);
+        return LV_RESULT_INVALID;
+    }
+
+    const uint8_t * in = file_buf;
+    uint8_t * out = img_data;
+    for(uint32_t y = 0; y < dsc->header.h; y++) {
+        decode_indexed_line(cf, dsc->palette, 0, 0, dsc->header.w, in, (lv_color32_t*)out);
+        in += dsc->header.stride;
+        out += stride;
+    }
+
+    dsc->header.stride = stride;
+    dsc->header.cf = LV_COLOR_FORMAT_ARGB8888;
+    dsc->img_data = img_data;
+    decoder_data->img_data = img_data; /*Free when decoder closes*/
+    lv_draw_buf_free(file_buf);
+#else
+    /*It needs to be read by get_area_cb later*/
+#endif
+    return LV_RESULT_OK;
+}
+
+static lv_result_t decode_rgb(lv_image_decoder_t * decoder, lv_image_decoder_dsc_t * dsc)
+{
```

Comment:
Shouldn't we use the cache here?

Reply:
True, I haven't considered cache. Will do it when `decode to RAM` is finished.